lib/hash_ring/ets.ex

```diff
@@ -287,83 +349,184 @@ defmodule ExHashRing.HashRing.ETS do
           pending_gcs
 
         {_stale_timer_ref, pending_gcs} ->
-          do_ring_gen_gc(table, ring_gen)
+          :ok = do_ring_gen_gc(state.previous_table, ring_gen)
+          :ok = do_ring_gen_gc(state.current_table, ring_gen)
           pending_gcs
       end
 
-    {:noreply, %{state | pending_gcs: pending_gcs}}
+    {:noreply, %__MODULE__{state | pending_gcs: pending_gcs}}
   end
 
-  defp rebuild(
-         %{nodes: nodes, name: name, table: table, ring_gen: ring_gen, overrides: overrides} =
-           state
-       ) do
-    new_ring_gen = ring_gen + 1
+  ## Private
 
-    :ets.insert(
-      table,
-      for {hash, node} <- Utils.gen_items(nodes) do
-        {{new_ring_gen, hash}, node}
-      end
-    )
+  @spec do_find_nodes_in_table(
+    key :: term(),
+    table :: :ets.tid(),
+    overrides :: override_map(),
+    gen :: generation(),
+    num_nodes :: non_neg_integer(),
+    num :: non_neg_integer()
+  ) :: {:ok, [binary]} | {:error, term}
+  defp do_find_nodes_in_table(_key, _table, _overrides, _gen, 0, _num) do
+    {:error, :invalid_ring}
+  end
+
+  defp do_find_nodes_in_table(_key, _table, _overrides, _gen, _num_nodes, 0) do
+    {:ok, []}
+  end
+
+  defp do_find_nodes_in_table(key, table, overrides, gen, num_nodes, num) when map_size(overrides) == 0 do
+    {:ok, do_find_nodes(table, gen, num_nodes, num, Utils.hash(key), [], 0)}
+  end
 
-    config =
-      if map_size(overrides) > 0 do
-        overrides =
-          overrides
-          |> Enum.filter(fn {_, values} -> length(values) > 0 end)
-          |> Map.new()
+  defp do_find_nodes_in_table(key, table, overrides, gen, num_nodes, num) do
+    {found, found_length} =
+      case overrides do
+        %{^key => overrides} ->
+          Utils.take_max(overrides, num)
 
-        {table, new_ring_gen, length(nodes), overrides}
-      else
-        {table, new_ring_gen, length(nodes)}
+        _ ->
+          {[], 0}
       end
 
-    Config.set(name, self(), config)
+    {:ok, do_find_nodes(table, gen, num_nodes, max(num - found_length, 0), Utils.hash(key), found, found_length)}
+  end
+
+  defp do_find_nodes(_table, _gen, _num_nodes, 0, _hash, found, _found_length) do
+    Enum.reverse(found)
+  end
 
-    schedule_ring_gen_gc(%{state | ring_gen: new_ring_gen}, ring_gen)
+  defp do_find_nodes(_table, _gen, num_nodes, _remaining, _hash, found, num_nodes) do
+    Enum.reverse(found)
   end
 
-  def schedule_ring_gen_gc(state, 0), do: state
+  defp do_find_nodes(table, gen, num_nodes, remaining, hash, found, found_length) do
+    {number, node} = find_next_highest_item(table, gen, num_nodes, hash)
+
+    if node in found do
+      do_find_nodes(
+        table,
+        gen,
+        num_nodes,
+        remaining,
+        number,
+        found,
+        found_length
+      )
+    else
+      do_find_nodes(
+        table,
+        gen,
+        num_nodes,
+        remaining - 1,
+        number,
+        [node | found],
+        found_length + 1
```

Comment:
I don't think `found_length` used in `do_find_nodes` at all??

Reply:
It's used in the escape head to detect that you've found the number of logical nodes available in the ring.

https://github.com/discord/ex_hash_ring/pull/10/files?file-filters%5B%5D=.ex#diff-8495c75f3d229f289b88e4080201c480R434